api_client.go

```diff
@@ -0,0 +1,1031 @@
+}
+
+// selectHeaderAccept join all accept types and return
+func selectHeaderAccept(accepts []string) string {
+    if len(accepts) == 0 {
+        return ""
+    }
+
+    if contains(accepts, "application/json") {
+        return "application/json"
+    }
+
+    return strings.Join(accepts, ",")
+}
+
+// contains is a case insensitive match, finding needle in a haystack
+func contains(haystack []string, needle string) bool {
+    for _, a := range haystack {
+        if strings.ToLower(a) == strings.ToLower(needle) {
+            return true
+        }
+    }
+    return false
+}
+
+// Verify optional parameters are of the correct type.
+func typeCheckParameter(obj interface{}, expected string, name string) error {
+    // Make sure there is an object.
+    if obj == nil {
+        return nil
+    }
+
+    // Check the type is as expected.
+    if reflect.TypeOf(obj).String() != expected {
+        return fmt.Errorf("Expected %s to be of type %s but received %s.", name, expected, reflect.TypeOf(obj).String())
+    }
+    return nil
+}
+
+// parameterToString convert interface{} parameters to string, using a delimiter if format is provided.
+func parameterToString(obj interface{}, collectionFormat string) string {
+    var delimiter string
+
+    switch collectionFormat {
+    case "pipes":
+        delimiter = "|"
+    case "ssv":
+        delimiter = " "
+    case "tsv":
+        delimiter = "\t"
+    case "csv":
+        delimiter = ","
+    }
+
+    if reflect.TypeOf(obj).Kind() == reflect.Slice {
+        return strings.Trim(strings.Replace(fmt.Sprint(obj), " ", delimiter, -1), "[]")
+    } else if t, ok := obj.(time.Time); ok {
+        return t.Format(time.RFC3339)
+    }
+
+    return fmt.Sprintf("%v", obj)
+}
+
+// helper for converting interface{} parameters to json strings
+func parameterToJson(obj interface{}) (string, error) {
+    jsonBuf, err := json.Marshal(obj)
+    if err != nil {
+        return "", err
+    }
+    return string(jsonBuf), err
+}
+
+// callAPI do the request.
+func (c *APIClient) callAPI(request *http.Request) (*http.Response, error) {
+    if c.cfg.Debug {
+        dump, err := httputil.DumpRequestOut(request, true)
+        if err != nil {
+            return nil, err
+        }
+        log.Printf("\n%s\n", string(dump))
+    }
+
+    resp, err := c.cfg.HTTPClient.Do(request)
+    if err != nil {
+        return resp, err
+    }
+
+    if c.cfg.Debug {
+        dump, err := httputil.DumpResponse(resp, true)
+        if err != nil {
+            return resp, err
+        }
+        log.Printf("\n%s\n", string(dump))
+    }
+    return resp, err
+}
+
+// Allow modification of underlying config for alternate implementations and testing
+// Caution: modifying the configuration while live can cause data races and potentially unwanted behavior
+func (c *APIClient) GetConfig() *Configuration {
+    return c.cfg
+}
+
+// prepareRequest build the request
+func (c *APIClient) prepareRequest(
+    ctx context.Context,
```

Comment:
This is a huge set of parameters?  Chance of organizing this better (i.e., group them in structs?)?

Reply:
this is auto-generated https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/go/client.mustache#L287